_posts/2024-03-13-bitcoin-inquisition-#45.md

```diff
@@ -11,17 +11,109 @@ status: upcoming
+
+Because this PR is for bitcoin inquisition, the
+[implementation](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1309)
+of `OP_INTERNALKEY` is remarkably simple. The taproot internal key is already
+extracted for use with verifying tapscript `OP_CHECKSIG(VERIFY|ADD)` with the
+key `0x01`, so it simply needs to be added to the stack.
+
+#### `OP_CHECKSIGFROMSTACK(VERIFY)`
+
+The implementation of CSFS(V) is split between the [main opcode
+processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).
 
+An earlier approach was to add a `SignatureChecker` that could carry the
+message rather than generating the message from the transaction, and give CSFS
+the exact semantics of existing sigops in their respective script types, but
+that approach had two problems: getting BIP340 signing into legacy/segwitv0
+scripts became intrusive, and the semantics of legacy/segwitv0 signature
+checking are not particularly desirable for a new opcode. As a result the
+current implementation uses a separate function specific to CSFS and gives
+modern Tapscript-style semantics. Only after changing code approaches did it
+become clear that ECDSA checking might not be desirable in these new sigops.
+The commit history still reflects the removal of ECDSA in a separate commit.
+
+Outside of these changes in `interpreter.cpp`, the only non-flag change in
+production code is to allow variable length input to BIP340
+signature-checking.
+
+### Testing
+
+Test data for both CHECKSIGFROMSTACK and INTERNALKEY has been added to
+[tx_valid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671)
+and
+[tx_invalid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491)
+for the trasnsaction test framework.
+[Here](https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd)
+are some scripts using `bitcoinjs-lib` and `noble-curves` that were used to
+generate the vectors.
+
+You can run a regtest node from this branch to use them yourself.
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+2. What does `OP_CHECKSIGFROMSTACKVERIFY` leave on the stack?
+   `OP_CHECKSIGFROMSTACK`?
+
+3. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?
+
+4. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?
+
+5. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend
+   in weight?
```

Comment:
Some more question ideas
- Did you run the tests? Is there any missing test coverage? Can you think of ways to extend the tests?
- How might one construct a vault (or one of the examples you mention above)?
- What concrete improvements to LN can be achieved this way?